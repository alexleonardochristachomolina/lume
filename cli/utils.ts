import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { isUrl } from "../core/utils/path.ts";
import { getConfigFile } from "../core/utils/lume_config.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(config?: string): Promise<Site> {
  let url: string | undefined;
  const watchInclude = [];

  if (config && isUrl(config)) {
    url = config;
  } else {
    const path = await getConfigFile(config);

    if (path) {
      watchInclude.push(path);
      url = toFileUrl(path).href;
    }
  }

  if (url) {
    log.info(`Loading config file <gray>${url}</gray>`);
    const mod = await import(url);
    const site = mod.default as Site | undefined;
    if (!(site instanceof Site)) {
      log.critical(
        `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
      );
      throw new Error("Site instance is not found");
    }

    site._data.configFile = url;
    site.options.watcher.include.push(...watchInclude);
    return site;
  }

  return lume();
}

/** Create a site intance and build it */
export async function buildSite(config?: string): Promise<Site> {
  const site = await createSite(config);

  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );

  return site;
}
