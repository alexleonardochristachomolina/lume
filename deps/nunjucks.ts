// @deno-types="npm:@types/nunjucks@3.2.6"
export { default } from "npm:nunjucks@3.2.4";
