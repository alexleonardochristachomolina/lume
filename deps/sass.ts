export * from "npm:sass@1.81.0";
