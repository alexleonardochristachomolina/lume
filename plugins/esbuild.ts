import {
  getPathAndExtension,
  isAbsolutePath,
  isUrl,
  normalizePath,
} from "../core/utils/path.ts";
import { merge } from "../core/utils/object.ts";
import { readDenoConfig } from "../core/utils/deno_config.ts";
import { log } from "../core/utils/log.ts";
import { readFile } from "../core/utils/read.ts";
import {
  build,
  BuildOptions,
  Metafile,
  OutputFile,
  stop,
} from "../deps/esbuild.ts";
import { extname, fromFileUrl, posix, toFileUrl } from "../deps/path.ts";
import { prepareAsset, saveAsset } from "./source_maps.ts";
import { Page } from "../core/file.ts";
import textLoader from "../core/loaders/text.ts";

import type Site from "../core/site.ts";
import type { DenoConfig } from "../core/utils/deno_config.ts";

export interface Options {
  /** The list of extensions this plugin applies to */
  extensions?: string[];

  /**
   * Global options for esm.sh CDN used to fetch NPM packages
   * @see https://esm.sh/#docs
   */
  esm?: EsmOptions;

  /**
   * The options for esbuild
   * @see https://esbuild.github.io/api/#general-options
   */
  options?: BuildOptions;
}

export interface EsmOptions {
  /** To include the ?dev option to all packages */
  dev?: boolean;

  /** Configure the cjs-exports option for each package */
  cjsExports?: Record<string, string[] | string>;

  /** Configure the deps for each package */
  deps?: Record<string, string[] | string>;
}

const denoConfig = await readDenoConfig();

// Default options
export const defaults: Options = {
  extensions: [".ts", ".js"],
  esm: {},
  options: {
    plugins: [],
    bundle: true,
    format: "esm",
    minify: true,
    keepNames: true,
    platform: "browser",
    target: "esnext",
    treeShaking: true,
    outdir: "./",
    outbase: ".",
  },
};

/**
 * A plugin to use esbuild in Lume
 * @see https://lume.land/plugins/esbuild/
 */
export function esbuild(userOptions?: Options) {
  const options = merge(defaults, userOptions);

  // Configure jsx automatically
  if (
    options.extensions.some((ext) =>
      ext.endsWith(".tsx") || ext.endsWith(".jsx")
    )
  ) {
    options.options = {
      ...buildJsxConfig(denoConfig?.config),
      ...options.options,
    };
  }

  // Sync the jsxDev option with esm.dev
  if (options.esm.dev) {
    options.options.jsxDev = true;
  } else if (options.options.jsxDev) {
    options.esm.dev = true;
  }

  return (site: Site) => {
    site.loadAssets(options.extensions);

    function resolve(path: string) {
      path = import.meta.resolve(path);
      const esmSpecifier = handleEsm(path, options.esm);

      return {
        path: esmSpecifier || path,
        namespace: "deno",
      };
    }

    site.hooks.addEsbuildPlugin = (plugin) => {
      options.options.plugins!.unshift(plugin);
    };

    site.addEventListener("beforeSave", stop);

    const basePath = options.options.absWorkingDir || site.src();

    const prefix = toFileUrl(site.src()).href;

    /** Run esbuild and returns the output files */
    const entryContent: Record<string, string> = {};

    async function runEsbuild(
      pages: Page[],
    ): Promise<[OutputFile[], Metafile, boolean]> {
      let enableAllSourceMaps = false;
      const entryPoints: { in: string; out: string }[] = [];

      pages.forEach((page) => {
        const { content, filename, enableSourceMap } = prepareAsset(site, page);
        if (enableSourceMap) {
          enableAllSourceMaps = true;
        }

        let outUri = getPathAndExtension(page.outputPath)[0];
        if (outUri.startsWith("/")) {
          outUri = outUri.slice(1); // This prevents Esbuild to generate urls with _.._/_.._/
        }

        entryPoints.push({ in: filename, out: outUri });
        entryContent[toFileUrl(filename).href] = content;
      });

      const buildOptions: BuildOptions = {
        ...options.options,
        write: false,
        metafile: true,
        absWorkingDir: basePath,
        entryPoints,
        sourcemap: enableAllSourceMaps ? "external" : undefined,
      };

      buildOptions.plugins!.push({
        name: "lume-loader",
        // deno-lint-ignore no-explicit-any
        setup(build: any) {
          build.onResolve({ filter: /.*/ }, (args: ResolveArguments) => {
            const { path, importer } = args;

            // Absolute url
            if (isUrl(path)) {
              return resolve(path);
            }

            // Resolve the relative url
            if (isUrl(importer) && path.match(/^[./]/)) {
              return resolve(new URL(path, importer).href);
            }

            // It's a npm package
            if (path.startsWith("npm:")) {
              return resolve(path);
            }

            if (!isUrl(path)) {
              return resolve(
                isAbsolutePath(path) ? toFileUrl(path).href : path,
              );
            }

            return resolve(path);
          });

          build.onLoad({ filter: /.*/ }, async (args: LoadArguments) => {
            let { path, namespace } = args;

            // It's one of the entry point files
            if (entryContent[path]) {
              return {
                contents: entryContent[path],
                loader: getLoader(path),
              };
            }

            // Read files from Lume
            if (namespace === "deno") {
              if (path.startsWith(prefix)) {
                const file = path.replace(prefix, "");
                const content = await site.getContent(file, textLoader);

                if (content) {
                  return {
                    contents: content,
                    loader: getLoader(path),
                  };
                }
              }
            }

            // Convert file:// urls to paths
            if (path.startsWith("file://")) {
              path = normalizePath(fromFileUrl(path));
            }

            // Read other files from the filesystem/url
            const content = await readFile(path);
            return {
              contents: content,
              loader: getLoader(path),
            };
          });
        },
      });

      const { outputFiles, metafile, warnings, errors } = await build(
        buildOptions,
      );

      if (errors.length) {
        log.error(`[esbuild plugin] ${errors.length} errors `);
      }

      if (warnings.length) {
        log.warn(
          `[esbuild plugin] ${warnings.length} warnings`,
        );
      }

      return [outputFiles || [], metafile!, enableAllSourceMaps];
    }

    site.process(options.extensions, async (pages, allPages) => {
      const [outputFiles, metafile, enableSourceMap] = await runEsbuild(pages);

      // Save the output code
      for (const [outputPath, output] of Object.entries(metafile.outputs)) {
        if (outputPath.endsWith(".map")) {
          continue;
        }

        const normalizedOutPath = normalizePath(outputPath);
        const outputFile = outputFiles.find((file) => {
          const relativeFilePath = normalizePath(
            normalizePath(file.path).replace(basePath, ""),
          );

          return relativeFilePath === normalizedOutPath;
        });

        if (!outputFile) {
          log.error(
            `[esbuild plugin] Could not match the metafile ${normalizedOutPath} to an output file.`,
          );
          continue;
        }

        // Replace .tsx and .jsx extensions with .js
        const content = options.options.bundle
          ? outputFile.text
          : resolveImports(
            outputFile.text,
            output.entryPoint
              ? relativePathFromUri(output.entryPoint, basePath)
              : outputPath,
            outputPath,
            basePath,
            options.esm,
            metafile,
          );

        // Get the associated source map
        const map = enableSourceMap
          ? outputFiles.find((f) => f.path === `${outputFile.path}.map`)
          : undefined;

        // Search the entry point of this output file
        let entryPoint: Page | undefined;

        if (output.entryPoint) {
          const outputRelativeEntryPoint = relativePathFromUri(
            output.entryPoint,
            basePath,
          );

          entryPoint = pages.find((page) =>
            page.sourcePath === outputRelativeEntryPoint
          );
        }

        // The page is a chunk
        if (!entryPoint) {
          const page = Page.create({ url: normalizedOutPath });
          saveAsset(site, page, content, map?.text);
          allPages.push(page);
          continue;
        }

        // The page is an entry point
        entryPoint.data.url = normalizedOutPath;
        saveAsset(site, entryPoint, content, map?.text);
      }
    });
  };
}

function getLoader(path: string) {
  const ext = extname(path).toLowerCase();

  switch (ext) {
    case ".ts":
    case ".mts":
      return "ts";
    case ".tsx":
      return "tsx";
    case ".jsx":
      return "jsx";
    case ".json":
      return "json";
    default:
      return "js";
  }
}

interface LoadArguments {
  path: string;
  namespace: string;
  suffix: string;
  pluginData: unknown;
}

interface ResolveArguments {
  path: string;
  importer: string;
  namespace: string;
  resolveDir: string;
  kind: string;
  pluginData: unknown;
}

function buildJsxConfig(config?: DenoConfig): BuildOptions | undefined {
  if (!config) {
    return;
  }

  const { compilerOptions } = config;

  if (compilerOptions?.jsxImportSource) {
    return {
      jsx: "automatic",
      jsxImportSource: compilerOptions.jsxImportSource,
    };
  }
}

function relativePathFromUri(uri: string, basePath?: string): string {
  if (uri.startsWith("deno:")) {
    uri = uri.slice("deno:".length);
  }

  if (uri.startsWith("file://")) {
    uri = fromFileUrl(uri);
  }

  return normalizePath(uri, basePath);
}

function resolveImports(
  source: string,
  sourcePath: string,
  outputPath: string,
  basePath: string,
  esm: EsmOptions,
  metafile: Metafile,
): string {
  source = source.replaceAll(
    /\bfrom\s*["']([^"']+)["']/g,
    (_, path) =>
      `from "${
        resolveImport(path, sourcePath, outputPath, basePath, esm, metafile)
      }"`,
  );

  source = source.replaceAll(
    /\bimport\s*["']([^"']+)["']/g,
    (_, path) =>
      `import "${
        resolveImport(path, sourcePath, outputPath, basePath, esm, metafile)
      }"`,
  );

  source = source.replaceAll(
    /\bimport\([\s\n]*["']([^"']+)["'](?=[\s\n]*[,)])/g,
    (_, path) =>
      `import("${
        resolveImport(path, sourcePath, outputPath, basePath, esm, metafile)
      }"`,
  );

  return source;
}

function resolveImport(
  importPath: string,
  sourcePath: string,
  outputPath: string,
  basePath: string,
  esm: EsmOptions,
  metafile: Metafile,
): string {
  if (importPath.startsWith(".") || importPath.startsWith("/")) {
    const sourcePathOfImport = posix.join(
      "/",
      posix.dirname(sourcePath),
      importPath,
    );

    const outputOfImport = Object.entries(metafile.outputs)
      .find(([_, output]) => {
        if (!output.entryPoint) {
          return false;
        }

        const outputRelativeEntryPoint = relativePathFromUri(
          output.entryPoint,
          basePath,
        );

        return outputRelativeEntryPoint === sourcePathOfImport;
      });

    if (!outputOfImport) {
      return importPath;
    }

    const outputPathOfImport = outputOfImport[0];
    const relativeOutputPathOfImport = posix.relative(
      posix.dirname(outputPath),
      outputPathOfImport,
    );

    return "./" + relativeOutputPathOfImport;
  }

  const resolved = import.meta.resolve(importPath);
  return handleEsm(resolved, esm) || resolved;
}

function handleEsm(path: string, options: EsmOptions): string | undefined {
  const match = path.match(/^(npm|jsr):(.*)$/);

  if (!match) {
    return;
  }

  const [, prefix, name] = match;
  const url = prefix === "npm"
    ? new URL(`https://esm.sh/${name}`)
    : new URL(`https://esm.sh/jsr/${name}`);

  if (options.dev) {
    url.searchParams.set("dev", "true");
  }

  // cjs exports
  const cjs_exports = options.cjsExports?.[name];

  if (cjs_exports) {
    url.searchParams.set(
      "cjs-exports",
      Array.isArray(cjs_exports) ? cjs_exports.join(",") : cjs_exports,
    );
  }

  // deps
  const deps = options.deps?.[name];

  if (deps) {
    url.searchParams.set(
      "deps",
      Array.isArray(deps) ? deps.join(",") : deps,
    );
  }

  return url.href;
}

export default esbuild;
