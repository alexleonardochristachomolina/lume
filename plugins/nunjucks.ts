import Nunjucks from "../deps/nunjucks.ts";
import loader from "../core/loaders/text.ts";
import { merge } from "../core/utils/object.ts";
import { normalizePath, resolveInclude } from "../core/utils/path.ts";
import { basename, join, posix } from "../deps/path.ts";

import type Site from "../core/site.ts";
import type { Engine, Helper, HelperOptions } from "../core/renderer.ts";
import type { ProxyComponents } from "../core/source.ts";

export interface Options {
  /** The list of extensions this plugin applies to */
  extensions?: string[];

  /** Optional sub-extension for page files */
  pageSubExtension?: string;

  /**
   * Custom includes path
   * @default `site.options.includes`
   */
  includes?: string;

  /**
   * Options passed to Nunjucks
   * @see https://mozilla.github.io/nunjucks/api.html#configure
   */
  options?: Nunjucks.ConfigureOptions;

  /** Plugins loaded by Nunjucks */
  plugins?: {
    [index: string]: Nunjucks.Extension;
  };
}

// Default options
export const defaults: Options = {
  extensions: [".njk"],
  options: {},
  plugins: {},
};

/** Template engine to render Nunjucks files */
export class NunjucksEngine implements Engine {
  // deno-lint-ignore no-explicit-any
  env: any;
  cache = new Map();
  basePath: string;
  includes: string;

  // deno-lint-ignore no-explicit-any
  constructor(env: any, basePath: string, includes: string) {
    this.env = env;
    this.basePath = basePath;
    this.includes = includes;
  }

  deleteCache(file: string): void {
    this.cache.delete(file);
    const filename = basename(file);

    // Remove the internal cache of nunjucks
    // deno-lint-ignore no-explicit-any
    this.env.loaders.forEach((fsLoader: any) => {
      Object.keys(fsLoader.cache).forEach((key) => {
        if (key.endsWith(filename)) {
          delete fsLoader.cache[key];
        }
      });
    });
  }

  render(
    content: string,
    data?: Record<string, unknown>,
    filename?: string,
  ): Promise<string> {
    if (!filename) {
      return new Promise((resolve, reject) => {
        this.env.renderString(content, data, (err: Error, result: string) => {
          if (err) {
            reject(err);
          } else {
            resolve(result);
          }
        });
      });
    }

    const template = this.getTemplate(content, filename);

    return new Promise((resolve, reject) => {
      template.render(data, (err: unknown, result: string) => {
        if (err) {
          reject(err);
        } else {
          resolve(result);
        }
      });
    });
  }

  renderComponent(
    content: string,
    data?: Record<string, unknown>,
    filename?: string,
  ): string {
    if (!filename) {
      return this.env.renderString(content, data);
    }

    const template = this.getTemplate(content, filename);
    return template.render(data);
  }

  getTemplate(content: string, filename: string) {
    if (!this.cache.has(filename)) {
      this.cache.set(
        filename,
        // @ts-ignore: The type definition of nunjucks is wrong
        Nunjucks.compile(content, this.env, join(this.basePath, filename)),
      );
    }

    return this.cache.get(filename)!;
  }

  addHelper(name: string, fn: Helper, options: HelperOptions) {
    switch (options.type) {
      case "tag": {
        const tag = createCustomTag(name, fn, options);
        this.env.addExtension(name, tag);
        return;
      }

      case "filter":
        if (options.async) {
          const filter = createAsyncFilter(fn);
          this.env.addFilter(name, filter, true);
          return;
        }

        // deno-lint-ignore no-explicit-any
        this.env.addFilter(name, function (this: any, ...args: unknown[]) {
          return fn.apply({ data: this.ctx }, args);
        });
    }
  }
}

class LumeLoader extends Nunjucks.Loader implements Nunjucks.ILoaderAsync {
  includes: string;

  constructor(private site: Site, includes: string) {
    super();
    this.includes = includes;
  }

  async: true = true;

  getSource(
    id: string,
    callback: Nunjucks.Callback<Error, Nunjucks.LoaderSource>,
  ) {
    const rootToRemove = this.site.src();
    let path = normalizePath(id, rootToRemove);

    if (path === normalizePath(id)) {
      path = resolveInclude(id, this.includes, undefined, rootToRemove);
    }

    this.site.getContent(path, loader).then((content) => {
      if (content) {
        callback(null, {
          src: content as string,
          path,
          noCache: false,
        });
        return;
      }

      callback(new Error(`Could not load ${path}`), null);
    });
  }

  override resolve(from: string, to: string): string {
    return posix.join(posix.dirname(from), to);
  }
}

/**
 * A plugin to use Nunjucks as a template engine
 * @see https://lume.land/plugins/nunjucks/
 */
export function nunjucks(userOptions?: Options) {
  return (site: Site) => {
    const options = merge(
      { ...defaults, includes: site.options.includes },
      userOptions,
    );

    const env = new Nunjucks.Environment(
      [new LumeLoader(site, options.includes)],
      options.options,
    );

    for (const [name, fn] of Object.entries(options.plugins)) {
      env.addExtension(name, fn);
    }

    site.hooks.addNunjucksPlugin = (name, fn) => {
      env.addExtension(name, fn);
    };

    const engine = new NunjucksEngine(env, site.src(), options.includes);

    // Ignore includes folder
    if (options.includes) {
      site.ignore(options.includes);
    }

    // Load the pages and register the engine
    site.loadPages(options.extensions, {
      loader,
      engine,
      pageSubExtension: options.pageSubExtension,
    });

    // Register the njk filter
    site.filter("njk", filter, true);

    // Register the component helper
    engine.addHelper("comp", (...args) => {
      const components = site.source.data.get("/")
        ?.[site.options.components.variable] as
          | ProxyComponents
          | undefined;
      const [content, name, options = {}] = args;
      delete options.__keywords;
      const props = { content, ...options };

      if (!components) {
        throw new Error(`Component "${name}" not found`);
      }

      const names = name.split(".") as string[];
      let component: ProxyComponents | undefined = components;

      while (names.length) {
        try {
          // @ts-ignore: `component` is defined or throw an error
          component = component[names.shift()];
        } catch {
          throw new Error(`Component "${name}" not found`);
        }
      }

      if (typeof component === "function") {
        return component(props);
      }

      throw new Error(`Component "${name}" not found`);
    }, {
      type: "tag",
      body: true,
    });

    function filter(
      string: string,
      data?: Record<string, unknown>,
    ): Promise<string> {
      return engine.render(string, { ...site.scopedData.get("/"), ...data });
    }
  };
}

/**
 * Create an asynchronous filter
 * by to adapting the Promise-based functions to callbacks used by Nunjucks
 * https://mozilla.github.io/nunjucks/api.html#custom-filters
 */
function createAsyncFilter(fn: Helper) {
  // deno-lint-ignore no-explicit-any
  return async function (this: any, ...args: unknown[]) {
    const cb = args.pop() as (err: unknown, result?: unknown) => void;

    try {
      const result = await fn.apply({ data: this.ctx }, args);
      cb(null, result);
    } catch (err) {
      cb(err);
    }
  };
}

/**
 * Create a tag extension
 * https://mozilla.github.io/nunjucks/api.html#custom-tags
 *
 * @param name The tag name
 * @param fn The function to render this tag
 * @param options The options to configure this tag
 */
function createCustomTag(name: string, fn: Helper, options: HelperOptions) {
  const tagExtension = {
    tags: [name],
    // @ts-ignore: There's no types for Nunjucks
    parse(parser, nodes) {
      const token = parser.nextToken();
      const args = parser.parseSignature(null, true);
      parser.advanceAfterBlockEnd(token.value);

      const extraArgs = [];

      if (options.body) {
        const body = parser.parseUntilBlocks(`end${name}`);
        extraArgs.push(body);
        parser.advanceAfterBlockEnd();
      }

      if (options.async) {
        return new nodes.CallExtensionAsync(
          tagExtension,
          "run",
          args,
          extraArgs,
        );
      }

      return new nodes.CallExtension(tagExtension, "run", args, extraArgs);
    },

    // deno-lint-ignore no-explicit-any
    run(context: any, ...args: any[]) {
      if (options.body) {
        const [body] = args.splice(
          options.async ? args.length - 2 : args.length - 1,
          1,
        );
        args.unshift(body());
      }

      if (!options.async) {
        const string = fn.apply({ data: context.ctx }, args);
        return new Nunjucks.runtime.SafeString(string);
      }

      const callback = args.pop();

      (fn.apply({ data: context.ctx }, args) as Promise<string>).then(
        (string: string) => {
          const result = new Nunjucks.runtime.SafeString(string);
          callback(null, result);
        },
      );
    },
  };

  return tagExtension;
}

export default nunjucks;

/** Extends Helpers interface */
declare global {
  namespace Lume {
    export interface Helpers {
      /** @see https://lume.land/plugins/nunjucks/ */
      njk: (string: string, data?: Record<string, unknown>) => Promise<string>;
    }
  }
}
