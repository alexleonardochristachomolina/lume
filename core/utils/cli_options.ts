import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { dev: "d", serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  const serveMode = cli.serve || cli._[0] === "cms";

  if (cli.port) {
    (options.server ||= {}).port = parseInt(cli.port);
  } else if (serveMode) {
    (options.server ||= {}).port = 3000;
  }

  let location: URL;

  if (cli.location) {
    location = new URL(cli.location);
  } else if (options.location && !serveMode) {
    location = options.location as URL;
  } else {
    location = new URL("http://localhost");
  }

  let port: number;

  if (cli.port) {
    port = parseInt(cli.port);
  } else if (location.port) {
    port = parseInt(location.port);
  } else if (options.server?.port) {
    port = options.server.port;
  } else {
    port = cli.serve ? 3000 : location.protocol === "https:" ? 443 : 80;
  }

  (options.server ||= {}).port = port;
  location.port = port.toString();
  options.location = location;

  if (cli.open) {
    (options.server ||= {}).open = cli.open;
  }

  return options;
}
