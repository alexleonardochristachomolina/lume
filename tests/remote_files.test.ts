import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import postcss from "../plugins/postcss.ts";
import sass from "../plugins/sass.ts";
import esbuild from "../plugins/esbuild.ts";
import nunjucks from "../plugins/nunjucks.ts";

Deno.test("render remote files", {
  sanitizeOps: false,
  sanitizeResources: false,
}, async (t) => {
  const site = getSite({
    src: "remote_files",
  });
  site.copy("asset.txt");
  site.use(postcss());
  site.use(sass());
  site.use(esbuild());
  site.use(nunjucks());

  const base = import.meta.resolve("./assets/remote_files/_remotes/");

  site.remoteFile("_includes/remote1.njk", new URL("./remote1.njk", base).href);
  site.remoteFile(
    "_includes/templates/remote-template2.njk",
    new URL("./remote-template2.njk", base).href,
  );
  site.remoteFile("asset.txt", new URL("./asset.txt", base).href);
  site.remoteFile("styles2.css", new URL("./styles2.css", base).href);
  site.remoteFile(
    "_includes/remote-style.css",
    new URL("./remote-style.css", base).href,
  );
  site.remoteFile(
    "other-remote-style.css",
    new URL("./other-remote-style.css", base).href,
  );
  site.remoteFile("_data.yml", new URL("./_data.yml", base).href);
  site.remoteFile("_includes/hello.js", new URL("./hello.js", base).href);
  site.remoteFile(
    "_includes/variables.scss",
    new URL("./variables.scss", base).href,
  );

  await build(site);
  await assertSiteSnapshot(t, site);
});
