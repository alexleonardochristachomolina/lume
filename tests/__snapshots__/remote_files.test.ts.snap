export const snapshot = {};

snapshot[`render remote files 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      loader: [AsyncFunction: toml],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.ts",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".page.js",
      loader: [AsyncFunction: module],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      engines: 0,
      ext: ".page.json",
      loader: [AsyncFunction: json],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".markdown",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      assetLoader: [AsyncFunction: text],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
      pageType: "asset",
    },
    {
      assetLoader: [AsyncFunction: text],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
      pageType: "asset",
    },
    {
      engines: 1,
      ext: ".vento",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      engines: 1,
      ext: ".vto",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      loader: [AsyncFunction: yaml],
      pageType: "page",
    },
    {
      assetLoader: [AsyncFunction: text],
      engines: undefined,
      ext: ".css",
      pageType: "asset",
    },
    {
      assetLoader: [AsyncFunction: text],
      engines: undefined,
      ext: ".scss",
      pageType: "asset",
    },
    {
      assetLoader: [AsyncFunction: text],
      engines: undefined,
      ext: ".sass",
      pageType: "asset",
    },
    {
      engines: 1,
      ext: ".njk",
      loader: [AsyncFunction: text],
      pageType: "page",
    },
  ],
  src: [
    "/",
    "/_data.yml",
    "/_includes",
    "/_includes/hello.js",
    "/_includes/name.js",
    "/_includes/remote-style.css",
    "/_includes/remote1.njk",
    "/_includes/templates",
    "/_includes/templates/local1.njk",
    "/_includes/templates/remote-template2.njk",
    "/_includes/variables.scss",
    "/_remotes",
    "/_remotes/_data.yml",
    "/_remotes/asset.txt",
    "/_remotes/hello.js",
    "/_remotes/other-remote-style.css",
    "/_remotes/remote-style.css",
    "/_remotes/remote-template2.njk",
    "/_remotes/remote1.njk",
    "/_remotes/styles2.css",
    "/_remotes/variables.scss",
    "/asset.txt",
    "/local1.md",
    "/other-remote-style.css",
    "/scripts.js",
    "/styles1.css",
    "/styles2.css",
    "/styles3.scss",
  ],
}
`;

snapshot[`render remote files 2`] = `
[
  {
    entry: "/asset.txt",
    flags: [
      "remote",
    ],
    outputPath: "/asset.txt",
  },
]
`;

snapshot[`render remote files 3`] = `
[
  {
    content: "<!DOCTYPE html>
<h1>Remote layout</h1>
<h2>Remote title</h2>

&lt;p&gt;Hello world&lt;/p&gt;


This is a local template

This is a remote template
",
    data: {
      basename: "local1",
      children: "<p>Hello world</p>
",
      content: "Hello world
",
      date: [],
      layout: "remote1.njk",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Remote title",
      url: "/local1/",
    },
    src: {
      asset: false,
      ext: ".md",
      path: "/local1",
      remote: undefined,
    },
  },
  {
    content: "/* Other remote style */
body {
  color: yellow;
}
",
    data: {
      basename: "other-remote-style",
      content: "/* Other remote style */
body {
  color: yellow;
}
",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Remote title",
      url: "/other-remote-style.css",
    },
    src: {
      asset: true,
      ext: ".css",
      path: "/other-remote-style",
      remote: "assets/remote_files/_remotes/other-remote-style.css",
    },
  },
  {
    content: 'var r=Object.defineProperty;var e=(t,m)=>r(t,"name",{value:m,configurable:!0});var l="Name";function o(){console.log("hello",l)}e(o,"default");o();
',
    data: {
      basename: "scripts",
      content: 'import hello from "./_includes/hello.js";

hello();
',
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Remote title",
      url: "/scripts.js",
    },
    src: {
      asset: true,
      ext: ".js",
      path: "/scripts",
      remote: undefined,
    },
  },
  {
    content: "/* Remote style */
body {
  color: red;
}
/* Other remote style */
body {
  color: yellow;
}
",
    data: {
      basename: "styles1",
      content: '@import "remote-style.css";
@import "./other-remote-style.css";
',
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Remote title",
      url: "/styles1.css",
    },
    src: {
      asset: true,
      ext: ".css",
      path: "/styles1",
      remote: undefined,
    },
  },
  {
    content: "body {
  color: blue;
}",
    data: {
      basename: "styles2",
      content: "body {
  color: blue;
}",
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Remote title",
      url: "/styles2.css",
    },
    src: {
      asset: true,
      ext: ".css",
      path: "/styles2",
      remote: "assets/remote_files/_remotes/styles2.css",
    },
  },
  {
    content: "body{color:blue}",
    data: {
      basename: "styles3",
      content: '@use "variables" as *;

body {
  color: \$color;
}
',
      date: [],
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
        "asset",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      title: "Remote title",
      url: "/styles3.css",
    },
    src: {
      asset: true,
      ext: ".scss",
      path: "/styles3",
      remote: undefined,
    },
  },
]
`;
